src/main/java/spiffe/provider/SpiffeTrustManager.java

```diff
@@ -0,0 +1,94 @@
+package spiffe.provider;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.X509ExtendedTrustManager;
+import java.net.Socket;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+
+import static spiffe.provider.CertificateUtils.checkSpiffeId;
+import static spiffe.provider.CertificateUtils.validate;
+
+/**
+ * This class implements a SPIFFE based TrustManager
+ *
+ */
+public class SpiffeTrustManager extends X509ExtendedTrustManager {
+
+
+    private final SpiffeIdManager spiffeIdManager;
+
+    SpiffeTrustManager() {
+        spiffeIdManager = SpiffeIdManager.getInstance();
+    }
+
+    /**
+     * Given the partial or complete certificate chain provided by the peer,
+     * build a certificate path to a trusted root and return if it can be validated
+     * and is trusted for client SSL authentication based on the authentication type.
+     *
+     * @param chain the peer certificate chain
+     * @param authType the authentication type based on the client certificate
+     * @throws CertificateException
+     */
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        checkPeer(chain);
+    }
+
+    /**
+     * Given the partial or complete certificate chain provided by the peer,
+     * build a certificate path to a trusted root and return if it can be validated
+     * and is trusted for server SSL authentication based on the authentication type.
+     *
+     * @param chain the peer certificate chain
+     * @param authType the key exchange algorithm used
+     * @throws CertificateException
+     */
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
```

Comment:
I assume this means that this provider will work for either server or client side? If it works for client side, then we'd probably want to be able to specify which SPIFFE IDs are expected on a per-client basis

Reply:
it could be for example having a ACL Manager that validates that a client is allowed to talk to a server based on the spiffeIDs: _acl.isAllowed(clientID, thisWorkload.getSpiffeId())_ , based on a list of access policies. Is that what you are thinking ?